Regression: integration test failed from HEAD

E0513 22:55:45.078788     957 controller.go:121] unable to sync kubernetes RO service: endpoints "kubernetes-ro" cannot be updated: the resource was updated to 266
E0513 22:55:45.080181     957 controller.go:121] unable to sync kubernetes RO service: endpoints "kubernetes-ro" cannot be updated: the resource was updated to 266
E0513 22:55:48.496436     957 reflector.go:165] watch of *api.Secret ended with: Stop requested
E0513 22:55:48.496531     957 reflector.go:165] watch of *api.ServiceAccount ended with: Stop requested
E0513 22:55:48.496610     957 reflector.go:165] watch of *api.Namespace ended with: Stop requested
E0513 22:55:48.496679     957 reflector.go:165] watch of *api.ServiceAccount ended with: Stop requested
E0513 22:55:48.496755     957 reflector.go:165] watch of *api.Secret ended with: Stop requested
E0513 22:55:48.496829     957 reflector.go:165] watch of *api.ServiceAccount ended with: Stop requested
W0513 22:55:49.950696     957 controller.go:228] Resetting endpoints for master service "kubernetes" to &{{ } {kubernetes  default    0001-01-01 00:00:00 +0000 UTC <nil> map[] map[]} [{[{10.0.0.22 <nil>}] [{ 6443 TCP}]}]}
W0513 22:55:50.032123     957 controller.go:228] Resetting endpoints for master service "kubernetes-ro" to &{{ } {kubernetes-ro  default    0001-01-01 00:00:00 +0000 UTC <nil> map[] map[]} [{[{10.0.0.22 <nil>}] [{ 7080 TCP}]}]}
W0513 22:55:50.032675     957 controller.go:228] Resetting endpoints for master service "kubernetes-ro" to &{{ } {kubernetes-ro  default    0001-01-01 00:00:00 +0000 UTC <nil> map[] map[]} [{[{10.0.0.22 <nil>}] [{ 7080 TCP}]}]}
W0513 22:55:50.033190     957 controller.go:228] Resetting endpoints for master service "kubernetes" to &{{ } {kubernetes  default    0001-01-01 00:00:00 +0000 UTC <nil> map[] map[]} [{[{10.0.0.22 <nil>}] [{ 6443 TCP}]}]}
W0513 22:55:50.113191     957 controller.go:228] Resetting endpoints for master service "kubernetes-ro" to &{{ } {kubernetes-ro  default    0001-01-01 00:00:00 +0000 UTC <nil> map[] map[]} [{[{10.0.0.22 <nil>}] [{ 7080 TCP}]}]}
W0513 22:55:50.172132     957 controller.go:228] Resetting endpoints for master service "kubernetes-ro" to &{{ } {kubernetes-ro  default    0001-01-01 00:00:00 +0000 UTC <nil> map[] map[]} [{[{10.0.0.22 <nil>}] [{ 7080 TCP}]}]}
W0513 22:55:50.193833     957 controller.go:228] Resetting endpoints for master service "kubernetes-ro" to &{{ } {kubernetes-ro  default    0001-01-01 00:00:00 +0000 UTC <nil> map[] map[]} [{[{10.0.0.22 <nil>}] [{ 7080 TCP}]}]}
E0513 22:55:50.394191     957 controller.go:92] unable to sync kubernetes service: endpoints "kubernetes" cannot be updated: the resource was updated to 279
E0513 22:55:50.395083     957 controller.go:121] unable to sync kubernetes RO service: endpoints "kubernetes-ro" cannot be updated: the resource was updated to 280
E0513 22:55:50.475196     957 controller.go:121] unable to sync kubernetes RO service: endpoints "kubernetes-ro" cannot be updated: the resource was updated to 280
E0513 22:55:50.476604     957 controller.go:121] unable to sync kubernetes RO service: endpoints "kubernetes-ro" cannot be updated: the resource was updated to 280
E0513 22:55:50.478775     957 controller.go:121] unable to sync kubernetes RO service: endpoints "kubernetes-ro" cannot be updated: the resource was updated to 280
W0513 22:55:51.538401     957 master.go:216] Portal net unspecified. Defaulting to 10.0.0.0/24.
I0513 22:55:51.538769     957 master.go:249] Will report 10.0.0.22 as public IP address.
E0513 22:56:00.274311     957 handlers.go:37] Unable to authenticate the request due to an error: Token has been invalidated
E0513 22:56:00.675185     957 handlers.go:37] Unable to authenticate the request due to an error: Token has been invalidated
E0513 22:56:00.877114     957 handlers.go:37] Unable to authenticate the request due to an error: Token has been invalidated
E0513 22:56:01.273530     957 handlers.go:37] Unable to authenticate the request due to an error: Token has been invalidated
E0513 22:56:08.076736     957 reflector.go:165] watch of *api.Secret ended with: Stop requested
E0513 22:56:08.076974     957 reflector.go:165] watch of *api.ServiceAccount ended with: Stop requested
E0513 22:56:08.077520     957 reflector.go:165] watch of *api.Namespace ended with: Stop requested
E0513 22:56:08.077759     957 reflector.go:165] watch of *api.ServiceAccount ended with: Stop requested
E0513 22:56:08.078025     957 reflector.go:165] watch of *api.Secret ended with: Stop requested
E0513 22:56:08.078269     957 reflector.go:165] watch of *api.ServiceAccount ended with: Stop requested
FAIL
FAIL	github.com/GoogleCloudPlatform/kubernetes/test/integration	111.315s

Here is the failing test: 
--- FAIL: TestPersistentVolumeClaimBinder (3.45s)
	persistent_volumes_test.go:106: Expected claim to be bound: &{TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:claim02 GenerateName: Namespace:default SelfLink:/api/v1beta1/persistentVolumeClaims/claim02?namespace=default UID:0e2d09c0-f9c3-11e4-96ec-02420a000016 ResourceVersion:172 CreationTimestamp:2015-05-13 22:54:47 +0000 UTC DeletionTimestamp:<nil> Labels:map[] Annotations:map[]} Spec:{AccessModes:[ReadOnlyMany ReadWriteOnce ReadWriteMany] Resources:{Limits:map[] Requests:map[storage:{Amount:5000000000 Format:BinarySI}]} VolumeName:} Status:{Phase:Pending AccessModes:[] Capacity:map[]}}

--- a/pkg/volumeclaimbinder/persistent_volume_index_test.go
+++ b/pkg/volumeclaimbinder/persistent_volume_index_test.go
@@ -112,11 +112,78 @@ func TestMatchVolume(t *testing.T) {
 			t.Errorf("Expected match but received nil volume for scenario: %s", name)
 		}
 		if scenario.expectedMatch != "" && volume != nil && string(volume.UID) != scenario.expectedMatch {
-			t.Errorf("Expected %s but got volume %s instead", scenario.expectedMatch, volume.UID)
+			t.Errorf("Expected %s but got volume %s in scenario %s", scenario.expectedMatch, volume.UID, name)
 		}
 		if scenario.expectedMatch == "" && volume != nil {
 			t.Errorf("Unexpected match for scenario: %s", name)
 		}
+	}
+}
+
+func TestMatchingWithBoundVolumes(t *testing.T) {
+	volumeIndex := NewPersistentVolumeOrderedIndex()
+	// two similar volumes, one is bound
+	pv1 := &api.PersistentVolume{
+		ObjectMeta: api.ObjectMeta{
+			UID:  "gce-pd-1",
+			Name: "gce001",
+		},
+		Spec: api.PersistentVolumeSpec{
+			Capacity: api.ResourceList{
+				api.ResourceName(api.ResourceStorage): resource.MustParse("1G"),
+			},
+			PersistentVolumeSource: api.PersistentVolumeSource{
+				GCEPersistentDisk: &api.GCEPersistentDiskVolumeSource{},
+			},
+			AccessModes: []api.PersistentVolumeAccessMode{api.ReadWriteOnce, api.ReadOnlyMany},
+			// this one we're pretending is already bound
+			ClaimRef: &api.ObjectReference{UID: "abc123"},
+		},
+	}
+
+	pv2 := &api.PersistentVolume{
+		ObjectMeta: api.ObjectMeta{
+			UID:  "gce-pd-2",
+			Name: "gce002",
+		},
+		Spec: api.PersistentVolumeSpec{
+			Capacity: api.ResourceList{
+				api.ResourceName(api.ResourceStorage): resource.MustParse("1G"),
+			},
+			PersistentVolumeSource: api.PersistentVolumeSource{
+				GCEPersistentDisk: &api.GCEPersistentDiskVolumeSource{},
+			},
+			AccessModes: []api.PersistentVolumeAccessMode{api.ReadWriteOnce, api.ReadOnlyMany},
+		},
+	}
+
+	volumeIndex.Add(pv1)
+	volumeIndex.Add(pv2)
+
+	claim := &api.PersistentVolumeClaim{
+		ObjectMeta: api.ObjectMeta{
+			Name:      "claim01",
+			Namespace: "myns",
+		},
+		Spec: api.PersistentVolumeClaimSpec{
+			AccessModes: []api.PersistentVolumeAccessMode{api.ReadOnlyMany, api.ReadWriteOnce},
+			Resources: api.ResourceRequirements{
+				Requests: api.ResourceList{
+					api.ResourceName(api.ResourceStorage): resource.MustParse("1G"),
+				},
+			},
+		},
+	}
+
+	volume, err := volumeIndex.FindBestMatchForClaim(claim)
+	if err != nil {
+		t.Fatalf("Unexpected error matching volume by claim: %v", err)
+	}
+	if volume == nil {
+		t.Fatalf("Unexpected nil volume.  Expected %s", pv2.Name)
+	}
+	if pv2.Name != volume.Name {
+		t.Errorf("Expected %s but got volume %s instead", pv2.Name, volume.Name)
 	}
 }
 
@@ -168,6 +235,26 @@ func createTestVolumes() []*api.PersistentVolume {
 					api.ReadWriteOnce,
 					api.ReadOnlyMany,
 				},
+			},
+		},
+		{
+			ObjectMeta: api.ObjectMeta{
+				UID:  "gce-pd-20",
+				Name: "gce004",
+			},
+			Spec: api.PersistentVolumeSpec{
+				Capacity: api.ResourceList{
+					api.ResourceName(api.ResourceStorage): resource.MustParse("20G"),
+				},
+				PersistentVolumeSource: api.PersistentVolumeSource{
+					GCEPersistentDisk: &api.GCEPersistentDiskVolumeSource{},
+				},
+				AccessModes: []api.PersistentVolumeAccessMode{
+					api.ReadWriteOnce,
+					api.ReadOnlyMany,
+				},
+				// this one we're pretending is already bound
+				ClaimRef: &api.ObjectReference{UID: "def456"},
 			},
 		},
 		{

--- a/pkg/volumeclaimbinder/types.go
+++ b/pkg/volumeclaimbinder/types.go
@@ -80,9 +80,18 @@ func (pvIndex *persistentVolumeOrderedIndex) Find(pv *api.PersistentVolume, matc
 		return nil, err
 	}
 
-	i := sort.Search(len(volumes), func(i int) bool { return matchPredicate(pv, volumes[i]) })
-	if i < len(volumes) {
-		return volumes[i], nil
+	// volumes are sorted by size but some may be bound.
+	// remove bound volumes for easy binary search by size
+	unboundVolumes := []*api.PersistentVolume{}
+	for _, v := range volumes {
+		if v.Spec.ClaimRef == nil {
+			unboundVolumes = append(unboundVolumes, v)
+		}
+	}
+
+	i := sort.Search(len(unboundVolumes), func(i int) bool { return matchPredicate(pv, unboundVolumes[i]) })
+	if i < len(unboundVolumes) {
+		return unboundVolumes[i], nil
 	}
 	return nil, nil
 }
@@ -97,8 +106,7 @@ func (pvIndex *persistentVolumeOrderedIndex) FindByAccessModesAndStorageCapacity
 			},
 		},
 	}
-
-	return pvIndex.Find(pv, filterBoundVolumes)
+	return pvIndex.Find(pv, matchStorageCapacity)
 }
 
 // FindBestMatchForClaim is a convenience method that finds a volume by the claim's AccessModes and requests for Storage
@@ -125,22 +133,9 @@ func (c byCapacity) Len() int {
 
 // matchStorageCapacity is a matchPredicate used to sort and find volumes
 func matchStorageCapacity(pvA, pvB *api.PersistentVolume) bool {
-	// skip already claimed volumes
-	if pvA.Spec.ClaimRef != nil {
-		return false
-	}
-
 	aQty := pvA.Spec.Capacity[api.ResourceStorage]
 	bQty := pvB.Spec.Capacity[api.ResourceStorage]
 	aSize := aQty.Value()
 	bSize := bQty.Value()
 	return aSize <= bSize
-}
-
-// filterBoundVolumes is a matchPredicate that filters bound volumes before comparing storage capacity
-func filterBoundVolumes(compareThis, toThis *api.PersistentVolume) bool {
-	if compareThis.Spec.ClaimRef != nil || toThis.Spec.ClaimRef != nil {
-		return false
-	}
-	return matchStorageCapacity(compareThis, toThis)
 }

